### ISSUE TITLE
Structure for skills?

### ISSUE THREAD

**Issue by bhaugen** [FreedomCoop/valuenetwork#156]
Somebody (I think @bum2 ) suggested that skills should have some hierarchy, which means parents. So I added a parent field to the form for creating resource types for skill suggestions.

But the rules in the code for parent resource types is that they need to have recipes, which means types of work cannot be parents. OOps. (But don't worry, that is easily changed, if we want.)

Anyway, what kinds of structure and classifications do we want for skills? And so we want those structures to appear on member profiles where they select their skills?

Classification possibilities in OCP right now include parents and facets, which you can see here:
https://testocp.freedomcoop.eu/accounting/resource-facets/

Let's discuss and decide.

**Comment by fosterlynn** [comment]
I think we should start with the faceted classification already available and then move to something else if this is not adequate.

I can help with the analysis of how to set it up for the skills.

**Comment by bum2** [comment]
Yes Lynn, i think it makes no sense to list all skills without any kind of grouping or hierachy... i'm used to do this kind of thing as a tree, normally with a drag'n'drop jquery to manage the MPTT tree in the admin side, and a chooser ready for mptt data (also jquery plugins) in the user side, to choose skills from the tree.
There's a django-mptt plugin to help manage this kind of data (i've tried long ago) but nowadays with just a 'parent' field we can build the structure.
I will search for the jquery plugins i used to manage the trees, and refresh my memory about mptt fields in django. (MPTT: Modified Preorder Tree Traversal, is a super fast way to read complex trees in SQL, it don't use recursivity).

**Comment by bhaugen** [comment]
MPTT is nice but will be one more dependency and require some database mods etc and some learning, unless you know how to handle it.

If we want hierarchy, I think the parent field and a recursive lookup(which already exists in the code) is fast enuf for the levels of hierarchy we are likely to ever have. But it won't get you drag-n-drop.

In the past, every time we tried either a single parent field or a single category field, it became too rigid pretty fast, which is why we moved to the faceted classification. Which seems difficult for people to configure, and so is a mixed blessing.

Anyway, we're not stuck on anything here. If you know and like MPTT and jquery drag-n-drop, go for it.

**Comment by fosterlynn** [comment]
Oops, I responded here #143 before I saw this.

**Comment by fosterlynn** [comment]
@bum2 sorry I should back off, forgot I already responded above, confused with the 2 issues. 

If you want to do a skills hierarchy, go for it.  It is conceptually easier in some ways.

Some initial design thoughts on how that would fit with what is there:

Do you think that the skills should still be resource types (types of work)?  I tentatively think so, that would enable people to plan and log work referencing that type of work, people would be notified of new work using their skills, etc., without changing the code in OCP.  Except there may have to be something to deal with traveling up or down in the hierarchy.

That makes me wonder if there is a good existing skills database out there that we could make use of?  Might short cut a long period of people having to request specific skills be added.

I also wonder if we should try to integrate that into the faceted classification, or not.  For example, would you want other facets, like "skill level" - "high", "medium", "low" (or whatever)? 

Maybe only the top level of resource types structured in a hierarchy are part of the faceted classification, and all the children inherit that as needed, but are part of a separate skills handling?

Somewhat random thoughts.

**Comment by bum2** [comment]
> Do you think that the skills should still be resource types (types of work)? I tentatively think so, that would enable people to plan and log work referencing that type of work, people would be notified of new work using their skills, etc., without changing the code in OCP. Except there may have to be something to deal with traveling up or down in the hierarchy.

Yes me too. Yes we will deal with traveling the tree, using the mptt functions... asap i'll try to make this Skills (types of work) as a hierachy, in a new branch, but let's finish the project-planning first! (pending to add the context agent into the exchange types, the patterns and into the resource types).

> I also wonder if we should try to integrate that into the faceted classification, or not. For example, would you want other facets, like "skill level" - "high", "medium", "low" (or whatever)? 

Is a good idea to keep the faceting possibility for this 'levels' or other categorisation we can need, then we can build or edit only one tree of skills with simple branches and leaves, the simpliest the best. These facets are working like 'categories' or 'common tags' to find or filter the records in a table, isn't it? 

If we want to set levels of knowledge or skill, is it needed a diferent skill record for each level? is there a way for a user to choose a skill and then choose the level of experience in that skill?

**Comment by fosterlynn** [comment]
> These facets are working like 'categories' or 'common tags' to find or filter the records in a table, isn't it? 

Basically - they are used to limit the resource type selections to the resource types that make sense on various pages.  https://en.wikipedia.org/wiki/Faceted_search

>If we want to set levels of knowledge or skill, is it needed a diferent skill record for each level? is there a way for a user to choose a skill and then choose the level of experience in that skill?

There is no way to record the level of experience for a skill.  Resource types are used for lots of things, so that would be probably too specific a thing to add (IMO).  You could do it now by creating a facet with facet values for the skill levels right now thought.  Those facet value would only be assigned to skills, and would work fine.  If the skills will be resource types (types of work), then this will be there already available.

**Comment by bum2** [comment]
good! then we only need to add the 'context_agent' FK field to the resource types and the 'parent' MPTT field to the skills only... is that possible? i see not, it's the same 'resource_type' model... ok, let's add the 'parent' field to the main 'resource_type'. I think we can try to mix the faceting already used with the branching-in-a-tree structure, and use one way or the other to filter resource types. That will be like having diferent trees one for 'types of work', another for 'categories' and another for 'project domain' (this facet is still not clear for me).
About the levels of experience, we will setup that using facets, but not now, just when it became a real need in a use case.

**Comment by bhaugen** [comment]
The relationship between an agent and a skill now has a field called "score":
https://github.com/FreedomCoop/valuenetwork/blob/master/valuenetwork/valueaccounting/models.py#L5669
which defined as "the quantity of contributions of this resource type from this agent". In terms of skills, it's usually how many hours a person has logged using this skill. 

We might be able to figure out how to register previous experience using that field, or add a new field.

**Comment by fosterlynn** [comment]
>and the 'parent' MPTT field to the skills only... is that possible? i see not, it's the same 'resource_type' model... ok, let's add the 'parent' field to the main 'resource_type'. 

Parent is there already.  (It is used to inherit recipes.)

**Comment by bum2** [comment]
Ok, we don't need to change the 'parent' field of EconomicResourceType into a tree mptt one because we can use the trees in the General app to show nested subcategories of Skills, and they are linked behind the scene with a real ocp resource_type with 'types of work' facet.
One of the reasons of adding this General app is to let the valuenetwork app code minimally touched. All kinds of needed trees can be easily managed with the general app and only some interface models in the work app connect to the ocp models.
I've worked out the first integration of general trees for choosing exchange types and resource types (general-tree branch, now integrating into project-planning), and as the skills are also resource_types it will be now easy to show them as a tree.
As I see it, skills are non-material resources, connected to services time and to the making of other material or non-material resources. From the General app point of view, skills (Job) are **verbs** that can connect resources, entities or places (general.Job and general.Relation have the same parent class: Art). 
From the NRP point of view, skills are only resource types like items, so the skills tree can be figured out using the 'services' branch of the work.Ocp_Nonmaterial_Type tree (not only used by the 'nonmaterial' facet but used also by the 'types of work' facet).

**Comment by bum2** [comment]
About the skills, I think it can be good to link them to the General job verbs tree, in a similar way like i did with other resource types. A interface model is proposed 'Ocp_Skill_Type' (first version in project-planning branch) that links facetvalues and resource 'types of work' with the general.job tree... When we create new skills in ocp, also the common general tree is updated (like with material or non-material 'artwork' types). That still not happens.
The 'services' branch makes no much sense for me as a non-material, it should be really the 'types of work' or skills 'model' (the third facet).
What is missing in the missing skill model is the main resource types related to each skill. The skills can be just verbs but needs to relate a material or non-material resource type to be defined... we still lack this way of doing skills around (i'll try to make a proposal using that new interface model)... meanwhile we can keep showing everywhere the plain list for everyone but that don't makes much sense for me in the long term.

**Comment by bhaugen** [comment]
> What is missing in the missing skill model is the main resource types related to each skill. The skills can be just verbs but needs to relate a material or non-material resource type to be defined.

The skills are now resource types. The skill suggestions become skill resource types. What am I missing in what you said?

**Comment by bum2** [comment]
Yes i know they are resource types. What i'm saying is a different concept or approach about skills: They are somehow different from other resource types and they don't have a special model in ocp, all the stuff about skills and resource types is mixed together in the same EconomicResource model. 
The way I see the skills is a like a double key-index: a verb and an name (artwork type, 'resource' type in ocp). The combination of both gives a skill, like in 'driving a truck' or 'repairing a truck', etc. Always a verb (action, skill) and usually related a resource type (can be not required). It can be a ManyToMany field between Job and Artwork_Type... but i want to let this subject for a future release and focus on finishing the exchanges and the new resource types (using the skills we have now).

**Comment by bhaugen** [comment]
> They are somehow different from other resource types and they don't have a special model in ocp, all the stuff about skills and resource types is mixed together in the same EconomicResource model.

Several EconomicResourceTypes have very different behaviors. Money, space, tools, types of work, products for sale, services that require payment, all have different behaviors.

The benefit of keeping them all together in the same EconomicResource and EconomicResourceType models is that they can then form recipes, plans, and resource flow graphs, and allow for accounting, value equations (that is, a contribution economy), and many other kinds of exchange and analysis without requiring tons of exception branches.

It is possible, sometimes, to say that a particular skill is usually or always applied to a particular resource type, but that is not always possible or useful. It would be possible to specify those connections if and when they are useful.

**Comment by bum2** [comment]
A new approach of setting new skills is proposed in the actual testing 'project-planning' branch, in the work/exchanges_all page. If you like and agree, we can use this form for other places where a skill is choosed, defined or edited, like in the user's profile. The skill suggestion workflow will keep running and can also be used for other types suggestions...

In the profile page is needed a list of choosed skills and a search system to add other skills to the personal set, and the button to create a new skill with the new form. The skills there can be created and initially set for the agent context, and the suggestion goes on. If the admin reviewing the skill suggestions sees it is a useful skill for other agents, can change the context and correct wording or parent type, etc. 

For all users having the same skill will be nice to let them all edit the skill description, and try to let them arrive to consensus about a description of the skill or job they all share. The name or verb editing can be reserved for the skill creator or for admins.

**Comment by bhaugen** [comment]
Ran into a bug when filtering the resource list:
https://testocp.freedomcoop.eu/work/agent/96/resources/
```
Exception Type:	KeyError
Exception Value:	
'name'
Exception Location:	/home/ocp/app/valuenetwork/work/forms.py in clean, line 1104
```

**Comment by bhaugen** [comment]
> In the profile page is needed a list of choosed skills and a search system to add other skills to the personal set, and the button to create a new skill with the new form. 

Does "is needed" mean this is not implemented yet?

**Comment by bum2** [comment]
> Does "is needed" mean this is not implemented yet?

Yes, you've implemented what we have in prod. I mean to do a list of choosed skills (without the non-choosed), no checkboxes, and then put the search autocomplete and the modal form for a new skill. Just to enhance what we have in the profile with the new widgets we have in the exchanges page, but keeping the same functionality and the suggestion workflow.

**Comment by fosterlynn** [comment]
Please ignore this if it is too different or hard to think about right now, but re. skills, I have been wondering if we just want to find a good open taxonomy of skills and use it.  I would be surprised if there isn't one or more already created.
